diff --git a/src/xpk/core/workload_decorators/tcpxo_decorator.py b/src/xpk/core/workload_decorators/tcpxo_decorator.py
index 771cafe..e455f2a 100644
--- a/src/xpk/core/workload_decorators/tcpxo_decorator.py
+++ b/src/xpk/core/workload_decorators/tcpxo_decorator.py
@@ -180,7 +180,7 @@ def update_gpu_containers(job_manifest):
     if 'nvidia.com/gpu' in container.get('resources', {}).get('limits', {}):
       container.setdefault('env', [])
       container['env'].append(
-          {'name': 'LD_LIBRARY_PATH', 'value': '/usr/local/nvidia/lib64'}
+          {'name': 'LD_LIBRARY_PATH', 'value': '/opt/nvidia/nccl/lib:/usr/local/nvidia/lib64'}
       )
       container['env'].append({
           'name': 'NCCL_FASTRAK_LLCM_DEVICE_DIRECTORY',
