diff --git a/pip_package/build.sh b/pip_package/build.sh
index ef62c432e..659e78956 100755
--- a/pip_package/build.sh
+++ b/pip_package/build.sh
@@ -89,7 +89,7 @@ bazel clean
 bazel build $@ ...
 if ! [[ $SKIP_TESTS ]]; then
   # Just test the core for the purposes of the pip package.
-  bazel test $@ lingvo/core/...
+  bazel test $@ lingvo/core/... --  -//lingvo/tasks/mt:model_test -//lingvo/core:saver_test
 fi

 DST_DIR="/tmp/lingvo/dist"
