diff --git a/nsys_recipe/lib/nvtx.py b/nsys_recipe/lib/nvtx.py
index 2470043..7abf892 100644
--- a/nsys_recipe/lib/nvtx.py
+++ b/nsys_recipe/lib/nvtx.py
@@ -161,6 +161,7 @@ def _compute_gpu_projection_df(
             "start": list(nvtx_gpu_start_dict.values()) + starts,
             "end": list(nvtx_gpu_end_dict.values()) + ends,
             "pid": nvtx_df.loc[list(nvtx_gpu_end_dict.keys()) + indices, "pid"],
+            "tid": nvtx_df.loc[list(nvtx_gpu_end_dict.keys()) + indices, "tid"],
         }
     )

diff --git a/nsys_recipe/recipes/nvtx_gpu_proj_trace/nvtx_gpu_proj_trace.py b/nsys_recipe/recipes/nvtx_gpu_proj_trace/nvtx_gpu_proj_trace.py
index cd60bf4..37e0d0d 100644
--- a/nsys_recipe/recipes/nvtx_gpu_proj_trace/nvtx_gpu_proj_trace.py
+++ b/nsys_recipe/recipes/nvtx_gpu_proj_trace/nvtx_gpu_proj_trace.py
@@ -96,6 +96,7 @@ class NvtxGpuProjTrace(recipe.Recipe):
             "start": "Start",
             "end": "End",
             "pid": "PID",
+            "tid": "TID",
             "stackLevel": "Stack Level",
             "childrenCount": "Children Count",
             "rangeId": "Range ID",
