diff --git a/xla/ef57.cc b/xla/ef57.cc
index d419f7810..9c526eeed 100644
--- a/xla/ef57.cc
+++ b/xla/ef57.cc
@@ -81,7 +81,7 @@ void ConvertF64ToEf57(absl::Span<const double> input,
     output.remove_prefix(kFloatsPerSseIteration);
   }
 #endif
-#if defined(XLA_HAS_ARM_NEON) && defined(XLA_HAS_ARM64)
+#if defined(XLA_HAS_ARM_NEON) && defined(XLA_HAS_ARM64) && defined(__clang__)
   constexpr int kDoublesPerNeonIteration = sizeof(float64x2_t) / sizeof(double);
   constexpr int kFloatsPerNeonIteration = sizeof(float32x2x2_t) / sizeof(float);
   while (input.size() >= kDoublesPerNeonIteration) {
