diff --git a/src/xpk/core/cluster.py b/src/xpk/core/cluster.py
index 1d9f234..56e16cd 100644
--- a/src/xpk/core/cluster.py
+++ b/src/xpk/core/cluster.py
@@ -118,6 +118,10 @@ def install_nccl_on_cluster(system: SystemCharacteristics) -> int:
       if system.gpu_config and system.gpu_config.nccl_installer
       else INSTALLER_NCCL_TCPXO
   )
+
+  if 'https://raw.githubusercontent.com' in nccl_installer:
+      nccl_installer = disable_nccl_install(nccl_installer)
+
   command = f'kubectl apply -f {nccl_installer}'
 
   return_code = run_command_with_updates(
@@ -133,6 +137,23 @@ def install_nccl_on_cluster(system: SystemCharacteristics) -> int:
   return 0
 
 
+def disable_nccl_install(nccl_installer: str) -> str:
+    import requests
+
+    yaml = requests.get(nccl_installer).text
+
+    yaml = yaml.replace(
+        "/scripts/container_entry.sh install --install-nccl",
+        "/scripts/container_entry.sh install"
+    )
+
+    manifest_path = "nccl-installer.yaml"
+    with open(manifest_path, "w") as f:
+        f.write(yaml_content)
+
+    return manifest_path
+
+
 def install_nri_on_cluster() -> int:
   """Install NRI Device Injector on the cluster.
 
