diff --git a/src/xpk/core/workload_decorators/tcpxo_decorator.py b/src/xpk/core/workload_decorators/tcpxo_decorator.py
index 93a2c0b..6a10aa8 100644
--- a/src/xpk/core/workload_decorators/tcpxo_decorator.py
+++ b/src/xpk/core/workload_decorators/tcpxo_decorator.py
@@ -146,7 +146,8 @@ def add_tcpxo_daemon_container(job_manifest):
           ' --num_hops=2 --num_nics=8 --uid= --alsologtostderr'
       ],
       'securityContext': {
-          'capabilities': {'add': ['NET_ADMIN', 'NET_BIND_SERVICE']}
+          'capabilities': {'add': ['NET_ADMIN', 'NET_BIND_SERVICE']},
+          'privileged': True,
       },
       'volumeMounts': [
           {'name': 'libraries', 'mountPath': '/usr/local/nvidia'},
@@ -161,11 +162,14 @@ def add_tcpxo_daemon_container(job_manifest):
 
 
 def update_gpu_containers(job_manifest):
+  job_manifest['spec']['template']['spec']['volumes'].append({
+      'name': 'tcpxo', 'hostPath': {'path': '/var/lib/tcpxo'}
+      })
   for container in job_manifest['spec']['template']['spec']['containers']:
     if 'nvidia.com/gpu' in container.get('resources', {}).get('limits', {}):
       container.setdefault('env', [])
       container['env'].append(
-          {'name': 'LD_LIBRARY_PATH', 'value': '/usr/local/nvidia/lib64'}
+          {'name': 'LD_LIBRARY_PATH', 'value': '/opt/tcpxo/lib64:/opt/nvidia/nccl/lib:/usr/local/nvidia/lib64'}
       )
       container['env'].append({
           'name': 'NCCL_FASTRAK_LLCM_DEVICE_DIRECTORY',
@@ -178,6 +182,9 @@ def update_gpu_containers(job_manifest):
       container['volumeMounts'].append(
           {'name': 'libraries', 'mountPath': '/usr/local/nvidia'}
       )
+      container['volumeMounts'].append(
+          {'name': 'tcpxo', 'mountPath': '/opt/tcpxo', 'readOnly': True}
+      )
       container['volumeMounts'].append(
           {'name': 'dshm', 'mountPath': '/dev/shm'}
       )
